File: backend_tapon/models/Order.js
const mongoose = require('mongoose');

const orderSchema = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  orderNumber: {
    type: String,
    required: true,
    unique: true
  },
  productType: {
    type: String,
    enum: ['nfc_card', 'review_card'],
    required: true
  },
  quantity: {
    type: Number,
    required: true,
    min: [1, 'Quantity must be at least 1']
  },
  totalAmount: {
    type: Number,
    required: true,
    min: [0, 'Total amount cannot be negative']
  },
  status: {
    type: String,
    enum: ['pending', 'processing', 'shipped', 'delivered', 'cancelled'],
    default: 'pending'
  },
  shippingAddress: {
    firstName: String,
    lastName: String,
    company: String,
    address1: String,
    address2: String,
    city: String,
    state: String,
    postalCode: String,
    country: String,
    phone: String
  },
  paymentStatus: {
    type: String,
    enum: ['pending', 'paid', 'failed', 'refunded'],
    default: 'pending'
  },
  stripePaymentIntentId: {
    type: String,
    default: null
  },
  trackingNumber: {
    type: String,
    default: null
  },
  estimatedDelivery: {
    type: Date,
    default: null
  },
  notes: {
    type: String,
    maxlength: [500, 'Notes cannot exceed 500 characters']
  }
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Virtual for order status
orderSchema.virtual('isDelivered').get(function() {
  return this.status === 'delivered';
});

orderSchema.virtual('isShipped').get(function() {
  return this.status === 'shipped';
});

orderSchema.virtual('isCancelled').get(function() {
  return this.status === 'cancelled';
});

// Indexes
orderSchema.index({ user: 1 });
orderSchema.index({ orderNumber: 1 });
orderSchema.index({ status: 1 });
orderSchema.index({ createdAt: -1 });

// Method to generate order number
orderSchema.statics.generateOrderNumber = function() {
  const timestamp = Date.now().toString();
  const random = Math.floor(Math.random() * 1000).toString().padStart(3, '0');
  return `TAP${timestamp}${random}`;
};

// Method to update status
orderSchema.methods.updateStatus = function(newStatus) {
  this.status = newStatus;
  if (newStatus === 'shipped') {
    this.estimatedDelivery = new Date(Date.now() + 7 * 24 * 60 * 60 * 1000); // 7 days
  }
  return this.save();
};

module.exports = mongoose.model('Order', orderSchema); 

File: backend_fastapi/app/models/order.py
from typing import Optional, List, Dict
from datetime import datetime
from beanie import Document, Indexed, PydanticObjectId
from pydantic import BaseModel, Field
from enum import Enum

class OrderStatus(str, Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    SHIPPED = "shipped"
    DELIVERED = "delivered"
    CANCELLED = "cancelled"

class PaymentStatus(str, Enum):
    PENDING = "pending"
    PAID = "paid"
    FAILED = "failed"
    REFUNDED = "refunded"

class ShippingAddress(BaseModel):
    firstName: Optional[str] = None
    lastName: Optional[str] = None
    company: Optional[str] = None
    address1: Optional[str] = None
    address2: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    postalCode: Optional[str] = None
    country: Optional[str] = None
    phone: Optional[str] = None

class OrderItem(BaseModel):
    productType: str
    quantity: int
    unitPrice: float
    # Add other item fields as needed based on frontend

class OrderNote(BaseModel):
    message: str
    addedBy: PydanticObjectId
    addedAt: datetime = Field(default_factory=datetime.utcnow)

class Order(Document):
    user: Indexed(PydanticObjectId)
    orderNumber: Indexed(str, unique=True)
    productType: str
    quantity: int
    items: List[OrderItem] = [] # Added to store line items
    totalAmount: float
    status: OrderStatus = OrderStatus.PENDING
    shippingAddress: ShippingAddress = ShippingAddress()
    paymentStatus: PaymentStatus = PaymentStatus.PENDING
    stripePaymentIntentId: Optional[str] = None
    trackingNumber: Optional[str] = None
    estimatedDelivery: Optional[datetime] = None
    notes: List[OrderNote] = [] # Changed from string to list based on controller usage
    
    paymentMethod: Optional[str] = None
    paymentTransactionId: Optional[str] = None
    paidAt: Optional[datetime] = None
    
    cancelledAt: Optional[datetime] = None
    cancellationReason: Optional[str] = None
    
    refundedAt: Optional[datetime] = None
    refundReason: Optional[str] = None
    refundAmount: Optional[float] = None

    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "orders"
